File: android/app/src/main/java/com/clavi/keyboard/WordPredictionEngine.kt
package com.clavi.keyboard

import android.content.Context
import java.io.IOException

/**
 * Word prediction engine (v1.5).
 *
 * Loads a per-language bigram TSV from assets (predict/{lang}.tsv):
 *   trigger_word<TAB>suggestion1<TAB>suggestion2<TAB>suggestion3
 *
 * Empty trigger (line starting with TAB) = default suggestions for start-of-sentence.
 *
 * Supports EN and UK; other languages silently return empty list.
 *
 * Usage:
 *   engine.predict("Hello how", Language.EN)  // → ["are", "much", "do"]
 *   engine.predict("", Language.UK)            // → ["Я", "Це", "Привіт"]
 */
class WordPredictionEngine(private val context: Context) {

    // Cache: lang key → bigram map
    private val cache = mutableMapOf<String, Map<String, List<String>>>()

    /**
     * Returns up to [maxResults] predicted next words given [textBefore] context.
     * Returns empty list if language has no prediction data or context gives no match.
     */
    fun predict(textBefore: String, language: Language, maxResults: Int = 3): List<String> {
        val assetKey = assetKey(language) ?: return emptyList()
        val dict = cache.getOrPut(assetKey) { loadDict(assetKey) }
        if (dict.isEmpty()) return emptyList()

        val lastWord = extractLastWord(textBefore).lowercase()

        // Exact match → suggestions for this trigger
        val exact = dict[lastWord]
        if (!exact.isNullOrEmpty()) return exact.take(maxResults)

        // No match for specific word → fall back to default context
        return (dict[""] ?: emptyList()).take(maxResults)
    }

    private fun assetKey(language: Language): String? = when (language) {
        Language.EN -> "en"
        Language.UK -> "uk"
        else        -> null   // no predictions for DE/FR/ES/etc. yet
    }

    private fun extractLastWord(text: String): String {
        val trimmed = text.trimEnd()
        if (trimmed.isEmpty()) return ""
        return trimmed.substringAfterLast(' ').substringAfterLast('\n')
    }

    private fun loadDict(lang: String): Map<String, List<String>> {
        return try {
            context.assets.open("predict/$lang.tsv")
                .bufferedReader()
                .lineSequence()
                .filter { it.isNotBlank() && !it.startsWith('#') }
                .mapNotNull { line ->
                    val parts = line.split('\t')
                    if (parts.isEmpty()) return@mapNotNull null
                    val trigger = parts[0]
                    val suggestions = parts.drop(1).filter { it.isNotBlank() }
                    if (suggestions.isEmpty()) return@mapNotNull null
                    trigger to suggestions
                }
                .toMap()
        } catch (_: IOException) {
            emptyMap()
        }
    }
}

File: ios/ClaviKeyboard/ClaviKeyboardExtension/WordPredictionEngine.swift
import Foundation

/**
 * Word prediction engine (v1.5) — mirrors WordPredictionEngine.kt exactly.
 *
 * Loads predict/{lang}.tsv from the extension bundle.
 * Format: trigger<TAB>sugg1<TAB>sugg2<TAB>sugg3
 * Empty trigger = default / sentence-start suggestions.
 *
 * Supports .en and .uk; other languages return [].
 */
struct WordPredictionEngine {

    // Cache: asset key → bigram dict (populated on first use per language)
    private var cache: [String: [String: [String]]] = [:]

    mutating func predict(_ textBefore: String, language: Language, maxResults: Int = 3) -> [String] {
        guard let assetKey = Self.assetKey(for: language) else { return [] }
        let dict = loadedDict(assetKey)
        if dict.isEmpty { return [] }

        let lastWord = extractLastWord(from: textBefore).lowercased()
        if let hits = dict[lastWord], !hits.isEmpty {
            return Array(hits.prefix(maxResults))
        }
        // Fall back to default context suggestions
        return Array((dict[""] ?? []).prefix(maxResults))
    }

    // MARK: - Private

    private static func assetKey(for language: Language) -> String? {
        switch language {
        case .en: return "en"
        case .uk: return "uk"
        default:  return nil
        }
    }

    private mutating func loadedDict(_ key: String) -> [String: [String]] {
        if let cached = cache[key] { return cached }
        let dict = Self.loadDict(key)
        cache[key] = dict
        return dict
    }

    private func extractLastWord(from text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .init(charactersIn: " \n"))
        if trimmed.isEmpty { return "" }
        return trimmed.components(separatedBy: CharacterSet(charactersIn: " \n")).last ?? ""
    }

    private static func loadDict(_ lang: String) -> [String: [String]] {
        guard let url = Bundle.main.url(forResource: lang, withExtension: "tsv",
                                        subdirectory: "predict"),
              let content = try? String(contentsOf: url, encoding: .utf8) else { return [:] }

        var result: [String: [String]] = [:]
        for line in content.components(separatedBy: "\n") {
            guard !line.hasPrefix("#"), !line.isEmpty else { continue }
            let parts = line.components(separatedBy: "\t")
            guard parts.count >= 2 else { continue }
            let trigger = parts[0]
            let suggestions = parts.dropFirst().filter { !$0.isEmpty }
            if !suggestions.isEmpty {
                result[trigger] = Array(suggestions)
            }
        }
        return result
    }
}
